modules/test/playwright/tests/frontend-data-set-admin-web/tests/data-set-admin/dataSetsPermissions.spec.ts

```diff
@@ -0,0 +1,734 @@
+					scope: 1,
+				},
+			],
+			roleType: 'regular',
+		});
+
+		createdRoleIds.push(dataSetUserRole.id);
+	});
+
+	await test.step('Create a new user', async () => {
+		userAccount = await apiHelpers.headlessAdminUser.postUserAccount();
+
+		userData[userAccount.alternateName] = {
+			name: userAccount.givenName,
+			password: 'test',
+			surname: userAccount.familyName,
+		};
+
+		createdUserIds.push(userAccount.id);
+	});
+
+	await test.step('Assign new role to user', async () => {
+		await apiHelpers.headlessAdminUser.postRoleUserAccountAssociation(
+			dataSetUserRole.id,
+			Number(userAccount.id)
+		);
+
+		apiHelpers.data.push({
+			id: `${dataSetUserRole.id}_${userAccount.id}`,
+			type: 'roleUserAccountAssociation',
+		});
+	});
+
+	// Enable Data Set roles through the UI since the Data Set Object created
+	// is given a random resource name (For example: com.liferay.object.model.ObjectDefinition#E0X3).
+
+	if (dataSetResourcePermissions) {
+		await test.step('Go to roles admin page', async () => {
+			await rolesPage.goto();
+		});
+
+		await test.step('Navigate to role edit page', async () => {
+			await page
+				.getByRole('link', {exact: true, name: dataSetUserRoleName})
+				.click();
+		});
+
+		await test.step('Navigate to "Define Permissions" > "Data Set" section', async () => {
+			await rolePage.definePermissionsLink.click();
+			await roleDefinePermissionsPage.searchInput.click();
+			await roleDefinePermissionsPage.searchInput.fill('Data Set');
+
+			await page
+				.getByRole('menuitem', {exact: true, name: 'Data Set'})
+				.click();
+		});
+
+		for (const dataSetResourcePermission of dataSetResourcePermissions) {
+			await test.step('Enable role checkboxes', async () => {
+				const dataSetRolesTable = page
+					.locator('.sheet-tertiary-title')
+					.getByText(dataSetResourcePermission.name, {exact: true})
+					.locator('~ .lfr-search-container');
+
+				for (const action of dataSetResourcePermission.actions) {
+					await dataSetRolesTable
+						.getByRole('row', {name: action})
+						.getByRole('checkbox')
+						.setChecked(true);
+				}
+			});
+		}
+
+		await test.step('Save roles', async () => {
+			await page.getByRole('button', {name: 'Save'}).click();
+
+			await waitForAlert(
+				page,
+				'Success:The role permissions were updated.'
+			);
+		});
+	}
+
+	await test.step('Do login with the new user', async () => {
+		await performLogout(page);
+		await performLogin(page, userAccount.alternateName);
+
+		loggedInAsAdmin = false;
+	});
+}
+
+test.beforeEach(async ({page}) => {
+
+	// Unsure why this happens, but sometimes this can start off as not
+	// signed in, so this attempts to log in again or else the headless
+	// requests will fail.
+
+	if (
+		await page
+			.getByRole('button', {
+				name: 'Sign In',
+			})
+			.isVisible()
+	) {
+		await test.step('Sign in as admin', async () => {
+			await performLogin(page, 'test');
+		});
+	}
```

Comment:
I've tried to rely on `loginTest()` fixture, but tests fail randomly (cannot create datasets nor roles due to lack of permissions) 

![image](https://github.com/user-attachments/assets/251be0e2-df92-4c2c-823e-c89d5584a01b)

I would say that Kevin's safeguard is fine for now, as tests run several times without issues.

Reply:
No way. I've added a `waitFor` to the `performLogout` and continue failing.
```js
await page.getByRole('button', {name: 'Sign In'}).waitFor({state: 'visible'});
```

It's weird, because the failing test shows the background page as logged in but no the main one
![image](https://github.com/user-attachments/assets/02f9675d-71fb-4613-b42d-d0481e4e0904)